### PR TITLE
add options for conversion quality management

### PR DESCRIPTION
**Pull request by jerpelea**
Many NuttX boards have limited resources and will have to use
compile time quality options to reduce resource usage

Signed-off-by: Alin Jerpelea <alin.jerpelea@sony.com>

### CODE REVIEW

**Comment by jerpelea** [comment]
@evpobr

**Review by evpobr — CHANGES_REQUESTED** [review]
It's a good idea, but the pull request needs to be improved. We will probably need a set of options for the build systems so that we can select the available converters, and this should be done for both Autotools and CMake.

**Comment by evpobr** [comment]
Look at #128.

**Comment by jerpelea** [comment]
@evpobr I can add the options for the build system for other OSes. In our case they are set in Kconfig for NuttX OS

**Comment by evpobr** [comment]
First we need to add ifdefs to libsamplerate's code and its buildsystems. I've started work in #128 but i have no time to complete it, i'll probably be very busy for the rest of the year. You can try if you wish.

**Comment by jerpelea** [comment]
@evpobr  it is alreaady done. please check my patch

**Review by evpobr — CHANGES_REQUESTED** [review]
There are build errors in Travis CI tests.

**Comment by jerpelea** [comment]
@evpobr local build and test was ok. I will fix the CI ones

**Comment by jerpelea** [comment]
@evpobr no more errors. please review

**Review by evpobr — CHANGES_REQUESTED** [review]
Sinc converter is disabled for Autotools build system, need options to set it on/off, enabled by default.

**Comment by jerpelea** [comment]
where should those options be located for autotools?

**Comment by evpobr** [comment]
This could help: https://autotools.io/autoconf/arguments.html.

**Comment by jerpelea** [comment]
@evpobr done. It seems to work. Please review

File: CMakeLists.txt [hunk 1]
```diff
@@ -27,6 +27,14 @@ option(LIBSAMPLERATE_EXAMPLES "Enable to generate examples" ${IS_ROOT_PROJECT})
 option(LIBSAMPLERATE_INSTALL "Enable to add install directives" ${IS_ROOT_PROJECT})
 option(LIBSAMPLERATE_COMPATIBLE_NAME "Use old dll name on Windows platform" OFF)
 
+option(LIBSAMPLERATE_WITH_SINC_FAST_CONVERTER "Enable Fastest Sinc Interpolator converter" ON)
```
Comment (evpobr) [inline CMakeLists.txt:30]: Maybe it is better to rename options to `LIBSAMPLERATE_ENABLE_SINC_ХХХ`? These names are more like the names of Autotools options. In addition, by tradition, Autotools` WITH ' means external dependencies.

File: src/samplerate.c [hunk 2]
```diff
@@ -496,11 +496,21 @@ psrc_set_converter (int converter_type, int channels, int *error)
 	SRC_STATE *state ;
 	switch (converter_type)
 	{
+#ifdef WITH_SINC_BEST_CONVERTER
```
Comment (evpobr) [inline src/samplerate.c:499]: And rename defines to something like `ENABLE_SINC_BEST_CONVERTER`.

**Review by evpobr — CHANGES_REQUESTED** [review]
Very good work, @jerpelea! After cleanup i've mentioned in my review, we will ask review from @SoapGentoo, i need to know that Autotools code is correct. And then i will merge it.

File: configure.ac [hunk 3]
```diff
@@ -213,6 +222,33 @@ AS_IF([test "x$enable_fftw" != "xno"], [
 			])
 	])
 
+dnl ====================================================================================
+dnl  Set Sinc Quality.
+
+AS_IF([test "x$enable_sinc_fast" != "xno"], [
+		dnl explicitly passed --enable-sinc-fast, hence error out loud and clearly
```
Comment (SoapGentoo) [inline configure.ac:229]: remove (just the comment)

File: configure.ac [hunk 4]
```diff
@@ -213,6 +222,33 @@ AS_IF([test "x$enable_fftw" != "xno"], [
...
+		AC_DEFINE_UNQUOTED([WITH_SINC_FAST_CONVERTER], [${enable_sinc_fast}], [Enable sinc fast converter.])
+	], [
+		dnl did not explicitly pass --enable-sinc-fast, relying on default automagic on
+		enable_sinc_fast="no"
+	])
+
+AS_IF([test "x$enable_sinc_medium" != "xno"], [
+		dnl explicitly passed --enable-sinc-medium, hence error out loud and clearly
```
Comment (SoapGentoo) [inline configure.ac:237]: remove

File: configure.ac [hunk 5]
```diff
@@ -213,6 +222,33 @@ AS_IF([test "x$enable_fftw" != "xno"], [
...
+		AC_DEFINE_UNQUOTED([WITH_SINC_MEDIUM_CONVERTER], [${enable_sinc_medium}], [Enable sinc balanced converter.])
+	], [
+		dnl did not explicitly pass --enable-sinc-medium, relying on default automagic on
+		enable_sinc_medium="no"
+	])
+
+AS_IF([test "x$enable_sinc_best" != "xno"], [
+		dnl explicitly passed --enable-sinc-best, hence error out loud and clearly
```
Comment (SoapGentoo) [inline configure.ac:245]: remove

File: src/src_sinc.c [hunk 6]
```diff
@@ -200,6 +206,7 @@ sinc_filter_new (int converter_type, int channels)
 	assert (converter_type == SRC_SINC_FASTEST ||
 		converter_type == SRC_SINC_MEDIUM_QUALITY ||
 		converter_type == SRC_SINC_BEST_QUALITY) ;
+
```
Comment (SoapGentoo) [inline src/src_sinc.c:209]: spurious whitespace change

**Review by SoapGentoo — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (jerpelea) [reply CMakeLists.txt]: done

**Comment by jerpelea** [comment]
@evpobr done

**Review by SoapGentoo — APPROVED** [review]

**Review by evpobr — APPROVED** [review]

**Comment by evpobr** [comment]
Thanks @jerpelea!

**Pull request merged**